utils/mil_ros_tools/online_bagger/nodes/online_bagger.py

```diff
@@ -0,0 +1,400 @@
+
+        for topic in self.topics:
+            self.topic_messages[topic] = sliceable_deque(deque())
+            self.subscriber_list.append([topic, False])
+
+        rospy.loginfo('failed topics: %s', self.subscriber_list)
+
+    def subscribe_loop(self):
+
+        """
+        Continue to subscribe until no topics fail
+        """
+
+        i = 0
+        while self.subscribe():
+            i = i + 1
+            self.subscribe
+            if not i % 1000:
+                rospy.loginfo('still subscribing!')
+
+    def subscribe(self):
+
+        """
+        Subscribe to the topics defined in the yaml configuration file
+
+        Function checks subscription status True/False of each topic
+        if True: topic has already been sucessfully subscribed to
+        if False: topic still needs to be subscribed to and
+        subscriber will be run.
+
+        Each element in self.subscriber list is a list [topic, Bool]
+        where the Bool tells the current status of the subscriber (sucess/failure).
+
+        Return number of topics that failed subscription
+        """
+
+        i = 0  # failed subscriptions
+
+        for topic in self.subscriber_list:
+            if not topic[1]:
+                msg_class = rostopic.get_topic_class(topic[0])
+                if msg_class[1] is None:
+                    i = i + 1
+                else:
+                    rospy.Subscriber(topic[0], msg_class[0],
+                                     lambda msg, _topic=topic[0]: self.bagger_callback(msg, _topic))
+
+                    topic[1] = True  # successful subscription
+
+        return i
+
+    def get_oldest_topic_time(self, topic):
+
+        """
+        Return oldest timestamp for a given topic as a rospy.Time type
+        """
+
+        return self.topic_messages[topic][0][0]
+
+    def get_newest_topic_time(self, topic):
+
+        """
+        Return newest timestamp for a given topic as a rospy.Time type
+        """
+
+        return self.topic_messages[topic][-1][0]
+
+    def get_topic_duration(self, topic):
+
+        """
+        Return current time duration of topic
+        """
+
+        return self.get_newest_topic_time(topic) - self.get_oldest_topic_time(topic)
+
+    def get_header_time(self, msg):
+
+        """
+        Retrieve header time if available
+        """
+
+        if hasattr(msg, 'header'):
+            return msg.header.stamp
+        else:
+            return rospy.get_rostime()
+
+    def get_ram_usage(self):
+
+        """
+        Return current ram usage in bytes
+        """
+
+        self.usage = resource.getrusage(resource.RUSAGE_SELF)
+
+        # getattr returns ram usage in kilobytes multiply by 1e3 to obtain megabytes
+        self.ram_usage = getattr(self.usage, 'ru_maxrss')*1e3
+
+        return self.ram_usage
+
+    def set_ram_limit(self, topic):
+
+        """
+        Limit ram usage by removing oldest messages
+        """
+
+        if self.get_ram_usage() > self.ram_limit:
```

Comment:
This won't really enforce the limit will it? If there's a topic that initially got small sized messages but now get's bigger messages the deleting will have no effect and lead to messages that don't need to be deleted being lost, or a topic that isn't published very frequently that contains big messages that hog all the RAM and rarely get's cleared.
New proposition: per topic RAM limits and using `sys.getsizeof(...)` to detect violations to the deque size. This will prevent losing small important data and make sure everything is as used as possible.